File: ios/CoinFactory.swift
//
//  CoinFactory.swift
//  coinswallet
//
//  Created by Tal Grynbaum on 30/06/2023.
//

import Foundation
import WalletCore

class CoinFactory {
  private static var coinMap = [String: (String) -> Coin]()
  
  static func registerCoin(name: String, constructor: @escaping (String) -> Coin) {
    coinMap[name] = constructor
  }
  
  static func createCoin(coinName: String, mnemonic: String) -> Coin {
    guard let constructor = coinMap[coinName.lowercased()] else {
      print("in create coin unsupported", coinName)
      fatalError("Unsupported coin: \(coinName)")
    }
    let coin =  constructor(mnemonic)
    print("after constructor coin: ", coin)
    return coin
  }
  class Coin {
    var wallet: HDWallet
    
    init(mnemonic: String) {
      guard let wallet = HDWallet(mnemonic: mnemonic, passphrase: "") else {
        fatalError("Failed to create HDWallet")
      }
      self.wallet = wallet
    }
    
    func getNewAddress(isTestNet:Bool) -> String {
      fatalError("Subclasses must override getNewAddress()")
    }
    
    func getPrivateKey(isTestNet:Bool) -> String {
      fatalError("Subclasses must override getPrivateKey()")
    }
    
    func signTransaction(rawData: String) -> String {
      fatalError("Subclasses must override signTransaction(rawData:)")
    }
    func getPublicKeyHash() -> String {
      return ""
    }
    func getExtendedPublicKey(isTestNet:Bool) -> String {
      return ""
    }
    func getExtendedPrivateKey(isTestNet:Bool) -> String {
      return ""
    }
    func getDeriveAddresses() -> NSMutableArray {
      return []
    }
    func getDeriveAddresses(isTestNet: Bool) -> NSMutableArray {
      return getDeriveAddresses()
    }
    func addCustomDerivation(derivePath:String, isTestNet:Bool) -> NSMutableDictionary {
      var params: NSMutableDictionary = [:]
      return params
    }
    // Mirrors RECEIVE_CHAIN / CHANGE_CHAIN / MAX_ADDRESSES_PER_CHAIN in
    // dok-wallet-blockchain-networks/service/bitcoinHdAddress.js.
    static let receiveChain = 0
    static let changeChain = 1
    static let maxDeriveRangeCount = 500

    // BIP44 account base path, e.g. "m/84'/0'/0'". The coin-type segment
    // differs per network (0' mainnet, 1' testnet), so overrides receive
    // isTestNet; this must stay in step with getAccountBasePath in
    // dok-wallet-blockchain-networks/service/bitcoinHdAddress.js. Coins that
    // support ranged derivation override this; others keep the empty default.
    func accountBasePath(isTestNet: Bool) -> String {
      return ""
    }
    // Derives `count` addresses on one BIP44 chain (0 = receive, 1 = change)
    // starting at `startIndex`, reusing each coin's addCustomDerivation so the
    // address type (bech32 / p2sh / legacy) stays coin-specific.
    //
    // Every step is a full BIP32 derivation, and this is reachable from JS
    // through the getDeriveAddressRange bridge method, so the range is
    // bounded: an unbounded count would block the caller and exhaust memory.
    // BIP44 defines exactly two chains, and the per-chain cap matches
    // MAX_ADDRESSES_PER_CHAIN in bitcoinHdAddress.js. The Int.max guard keeps
    // the range expression below from trapping on overflow. Invalid input
    // yields no addresses, as before.
    func getDeriveAddressRange(chainIndex: Int, startIndex: Int, count: Int, isTestNet: Bool) -> NSMutableArray {
      let result = NSMutableArray()
      let basePath = accountBasePath(isTestNet: isTestNet)
      if basePath.isEmpty
        || count <= 0
        || count > Coin.maxDeriveRangeCount
        || startIndex < 0
        || (chainIndex != Coin.receiveChain && chainIndex != Coin.changeChain)
        || startIndex > Int.max - count {
        return result
      }
      for i in startIndex..<(startIndex + count) {
        result.add(addCustomDerivation(derivePath: "\(basePath)/\(chainIndex)/\(i)", isTestNet: isTestNet))
      }
      return result
    }
  }
}

File: android/app/src/main/java/com/coinswallet/CoinFactory.java
// CoinFactory.java
package com.coinswallet;

import wallet.core.jni.*;
import java.util.*;
import java.util.function.Function;
import java.util.function.Supplier;
import android.util.Log;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableArray;

public class CoinFactory {

    // private static final Map<String, Supplier<Coin>> coinMap = new HashMap<>();
    private static final Map<String, Function<String, Coin>> coinMap = new HashMap<>();

    public static void registerCoin(String name, Function<String, Coin> constructor) {
        coinMap.put(name, constructor);
    }

    public static Coin createCoin(String coinName, String mnemonic) {
        Function<String, Coin> constructor = coinMap.get(coinName.toLowerCase());
        if (constructor == null) {
            throw new IllegalArgumentException("Unsupported coin: " + coinName);
        }
        return constructor.apply(mnemonic);
    }

    public static abstract class Coin {
        protected HDWallet wallet;

        protected Coin(String mnemonic) {
            this.wallet = new HDWallet(mnemonic, "");
        }

        public static String toHex(byte[] bytes) {
            StringBuilder sb = new StringBuilder();
            for (byte b : bytes) {
                sb.append(String.format("%02X", b));
            }
            return sb.toString();
        }

        public abstract String getNewAddress(Boolean isTestNet);

        public abstract String getPrivateKey(Boolean isTestNet);
        public String getPublicKeyHex(){
            return null;
        };
        public String getExtendedPublicKey(Boolean isTestNet) {return null; };

        public String getExtendedPrivateKey(Boolean isTestNet) {return null; };

        public ReadableArray getDeriveAddresses(){
            return null;
        };
        public ReadableArray getDeriveAddresses(Boolean isTestNet){
            return getDeriveAddresses();
        };
        public ReadableMap addCustomDerivation(String derivePath, Boolean isTestNet){
            return null;
        };
        // Mirrors RECEIVE_CHAIN / CHANGE_CHAIN / MAX_ADDRESSES_PER_CHAIN in
        // dok-wallet-blockchain-networks/service/bitcoinHdAddress.js.
        protected static final int RECEIVE_CHAIN = 0;
        protected static final int CHANGE_CHAIN = 1;
        protected static final int MAX_DERIVE_RANGE_COUNT = 500;

        // BIP44 account base path, e.g. "m/84'/0'/0'". The coin-type segment
        // differs per network (0' mainnet, 1' testnet), so overrides receive
        // isTestNet; this must stay in step with getAccountBasePath in
        // dok-wallet-blockchain-networks/service/bitcoinHdAddress.js. Coins that
        // support ranged derivation override this; others keep the empty default.
        public String accountBasePath(Boolean isTestNet) {
            return "";
        }
        // Appends `count` addresses of one BIP44 chain (0 = receive,
        // 1 = change) starting at `startIndex`, reusing each coin's
        // addCustomDerivation so the address type stays coin-specific.
        //
        // Every step is a full BIP32 derivation, and this is reachable from JS
        // through the getDeriveAddressRange @ReactMethod, so the range is
        // bounded: an unbounded count would block the bridge thread and
        // exhaust memory. BIP44 defines exactly two chains, and the per-chain
        // cap matches MAX_ADDRESSES_PER_CHAIN in bitcoinHdAddress.js. Invalid
        // input yields no addresses, as before.
        protected void appendDeriveAddressRange(WritableArray target, int chainIndex, int startIndex, int count, Boolean isTestNet) {
            String basePath = accountBasePath(isTestNet);
            if (basePath.isEmpty()
                    || count <= 0
                    || count > MAX_DERIVE_RANGE_COUNT
                    || startIndex < 0
                    || (chainIndex != RECEIVE_CHAIN && chainIndex != CHANGE_CHAIN)
                    || startIndex > Integer.MAX_VALUE - count) {
                return;
            }
            for (int i = startIndex; i < startIndex + count; i++) {
                target.pushMap(addCustomDerivation(basePath + "/" + chainIndex + "/" + i, isTestNet));
            }
        }
        public ReadableArray getDeriveAddressRange(int chainIndex, int startIndex, int count, Boolean isTestNet) {
            WritableArray result = Arguments.createArray();
            appendDeriveAddressRange(result, chainIndex, startIndex, count, isTestNet);
            return result;
        }
        public abstract String signTransaction(String rawData);
    }
}
